Some fixes to selector serialization re: namespaces and universal selector

jyc opened the pull request:
- Fix eliding default namespace when serializing
- Fix shortest serialization property when namespace prefix is `*|` and there is no default namespace
- Omit universal selector when serializing to match `cssom/serialize-namespaced-type-selectors` (again so we get the shortest serialization)

<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes is part of a series to fix #17182

<!-- Either: -->

I'd like to land #17501 first, because it allows some tests for this to work.

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/17537)
<!-- Reviewable:end -->

jyc committed:
Replace Namespace components which map to the default namespace with DefaultNamespace.

It seems we don't need to preserve the original prefix name, and this
conveniently leads to_css to now implement CSSOM's requirement to elide
the namespace prefix during serialization when it maps to the default
namespace.

https://drafts.csswg.org/cssom/#serialize-a-simple-selector

emilio commented:
@bors-servo try

@jyc this looks good to me, is this waiting for anything in particular? Does it have tests?

bors-servo commented:
:hourglass: Trying commit eaa1d40c0ead51a98216c144619ba448c5894685 with merge 9e0d601178f6b39d92c9a71bf94b26c592f6c31b...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/4767)

jyc commented:
Hi @emilio! Thanks, sorry for not being clear; was planning to run it to see what tests changed. It looks like it failed due to something in the infrastructure--

@bors-servo retry

bors-servo commented:
:hourglass: Trying commit eaa1d40c0ead51a98216c144619ba448c5894685 with merge 5b7fc90ee35cf97731e76548a807a16cd9340f2c...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/7729), [arm32](http://build.servo.org/builders/arm32/builds/7948), [arm64](http://build.servo.org/builders/arm64/builds/7984), [linux-dev](http://build.servo.org/builders/linux-dev/builds/8074), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/4828), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/4861), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/8062), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/541), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/538), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/4769), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/4825), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/539), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/544), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3307)
State: approved= try=True

jyc commented:
@bors-servo: retry now that #17501 is merged

jyc committed:
Elide the explicit any namespace prefix when lacking default namespace.

Once again it seems we don't need to preserve the original prefix name,
and this lets to_css serialize to the shortest form when there is no
default namespace and the *| prefix is used.

Selectors § 6.1.1 says:

    Element type selectors that have no namespace component (no namespace
    separator) represent elements without regard to the element's namespace
    (equivalent to "*|") unless a default namespace has been declared for
    namespaced selectors (e.g. in CSS, in the style sheet). If a default
    namespace has been declared, such selectors will represent only elements
    in the default namespace.

Then if there is no default namespace, *| (which we write as
QNamePrefix::ExplicitAnyNamespace) is equivalent to what we write as
QNamePrefix::ImplicitAnyNamespace; the latter has a shorter
serialization, so we should use that.

jdm commented:
```
     Running /home/travis/build/servo/servo/target/debug/deps/selectors-a25afd58e67e341d

running 11 tests

test parser::tests::test_empty ... ok

test parser::tests::test_empty_pseudo_iter ... ok

test parser::tests::test_pseudo_iter ... ok

test bloom::create_and_insert_some_stuff ... ok

test size_of_tests::size_of_component ... ok

test size_of_tests::size_of_pseudo_class ... ok

test size_of_tests::size_of_pseudo_element ... ok

test size_of_tests::size_of_selector ... ok

test parser::tests::visitor ... ok

test parser::tests::test_parsing ... FAILED

test parser::tests::test_universal ... FAILED
```

jyc commented:
The tests used a parser with no default namespace but asserted that *| should be preserved in the output, which is contrary to the tests in `serialize-namespaced-type-selectors.htm`. Fixed those (and added tests for when there is a default namespace), and fixing the merge conflict now.

jyc commented:
@bors-servo: retry

jyc commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 4269ce377efd8eaa806f11744088377c3d28f473 with merge c09874535adabf9a4592ee877405c24d0b656918...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/4953)

jdm commented:
```
  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with any namespace

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Universal selector in any namespace

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with any namespace followed by class

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with any namespace followed by id

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with any namespace followed by pseudo class

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with any namespace followed by pseudo element

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with any namespace followed by attribute selector

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with namespace equal to default namespace

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Universal selector with namespace equal to default namespace

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with namespace equal to default namespace followed by class

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with namespace equal to default namespace followed by id

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with namespace equal to default namespace followed by pseudo class

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with namespace equal to default namespace followed by pseudo element

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with namespace equal to default namespace followed by attribute selector
```

jyc commented:
Thanks again! Also, I think it should be pretty straightforward to fix the remaining failures in that file. Working on it!

jyc commented:
Created a spec issue at https://github.com/w3c/csswg-drafts/issues/1606 regarding what looks like an error in the spec vs. the web-platform test (which browsers pass).

jyc committed:
Omit serializing the universal selector when possible.

If omitting the universal selector in the serialization is possible, we
should do it so we obtain a shorter serialization (to match the behavior
asserted in cssom/serialize-namespaced-type-selectors.html).

For example, if someone writes *|*::before and there is no default
namespace, we should serialize to ::before; however, if there is a
default namespace, we should serialize to *|*::before.
(This is the test case "Universal selector in any namespace followed by
pseudo element).

This matches the behavior implemented by WebKit; that one case in
particular isn't implemented by Gecko, but other cases where the
universal selector should be elided are implemented by Gecko but were not
previously by Servo.

jyc commented:
@bors-servo: retry

emilio commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit a6a37f33f0f6958c31c9b03ce8cb5f78020e762e with merge 0e5c718848780be5121a37098e71bd15f1ddc90c...

emilio commented:
@jyc when there's a new commit you need to use `try` instead of `retry` :P

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/8263)

emilio commented:
```
	thread 'parser::tests::test_parsing' panicked at 'index out of bounds: the len is 0 but the index is 0', /home/servo/buildbot/slave/linux-dev/build/components/selectors/parser.rs:776:59
```

jyc commented:
@emilio Thanks! Makes sense.

jyc commented:
@bors-servo retry

KiChjang commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit b19434b99239761177179d7fb19d5ee43d18a7f7 with merge df1a7f6634d390838ec318e71e84b1a00f4c7bbe...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/7932), [arm32](http://build.servo.org/builders/arm32/builds/8154), [arm64](http://build.servo.org/builders/arm64/builds/8188), [linux-dev](http://build.servo.org/builders/linux-dev/builds/8274), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5034), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5067), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/8268), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/746), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/744), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/4971), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5029), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/747), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/752), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3512)
State: approved= try=True

jyc commented:
Problem was the fake combinators we have for pseudo element selectors. Should be good now!

@bors-servo r?emilio

KiChjang commented:
r? @emilio

emilio commented on components/selectors/parser.rs:
```diff
@@ -763,19 +763,99 @@ impl<Impl: SelectorImpl> ToCss for Selector<Impl> {
-            for item in compound.iter() {
-                item.to_css(dest)?;
+
+            // https://www.w3.org/TR/cssom-1/#serializing-selectors
```
nit: We usually link to `drafts.csswg.org`.

emilio commented on components/selectors/parser.rs:
```diff
@@ -763,19 +763,99 @@ impl<Impl: SelectorImpl> ToCss for Selector<Impl> {
+            //
+            // If we are in this case, we continue to the next iteration of the
+            // `for compound in compound_selectors` loop.
+            let namespace = if compound.len() > 0 {
```
nit: `if !compound.is_empty()`.

emilio commented on components/selectors/parser.rs:
```diff
@@ -763,19 +763,99 @@ impl<Impl: SelectorImpl> ToCss for Selector<Impl> {
+            // If we are in this case, we continue to the next iteration of the
+            // `for compound in compound_selectors` loop.
+            let namespace = if compound.len() > 0 {
+                let (namespace, first_non_namespace) = match &compound[0] {
```
nit: Why the reference? I think it'd be clearer to do:

```rust
let has_namespace = matches!(compound[0], Component::ExplicitAnyNamespace | ...);
```

And derive the rest from that.

jyc replied:
Hm, I think it is clearer to keep the logic as close to where it is used as possible. Also, we don't quite want `has_namespace`, I think we instead want something like `can_elide_namespace`. Going to include proposed change, though.

emilio commented on components/selectors/parser.rs:
```diff
@@ -763,19 +763,99 @@ impl<Impl: SelectorImpl> ToCss for Selector<Impl> {
+            // in cssom/serialize-namespaced-type-selectors.html, which the
+            // following code tries to match.
+            for simple in compound.iter() {
+                if let &Component::ExplicitUniversalType = simple {
```
nit: we usually do:

```rust
if let Component::ExplicitUniversalType = *simple {
```

or,

```rust
if matches!(*simple, Component::ExplicitUniversalType) {
```

jyc replied:
Didn't know about the `matches!` macro, thanks!

emilio commented on components/selectors/parser.rs:
```diff
@@ -763,19 +763,99 @@ impl<Impl: SelectorImpl> ToCss for Selector<Impl> {
+                    // compound selector, so we don't have to worry about the
+                    // real namespace being in a different `compound`.
+                    match namespace {
+                        Some(&Component::DefaultNamespace(_)) | None => continue,
```
nit: You can also use `matches!` here if you prefer, and avoid the `_ => ()` thing.

emilio commented on components/selectors/parser.rs:
```diff
@@ -1729,19 +1826,44 @@ pub mod tests {
+        input: &'i str,
+        parser: &DummyParser,
+        expected: Option<&'a str>
+    ) -> Result<SelectorList<DummySelectorImpl>, ParseError<'i, SelectorParseError<'i, ()>>>
```
nit: I think the brace should be in the previous line per https://github.com/rust-lang-nursery/fmt-rfcs

emilio commented on components/selectors/parser.rs:
```diff
@@ -2060,7 +2211,10 @@ pub mod tests {
     fn test_universal() {
-        let selector = &parse("*|*::before").unwrap().0[0];
+        let selector = &parse_ns("*|*::before", &DummyParser {
+            default_ns: Some(DummyAtom::from("https://mozilla.org")),
```
nit: Maybe having a `DummyParser::default()` or something like that would be nice, to avoid some duplication here.

jyc replied:
Good idea!

emilio reviewed (commented):
Looks good! Mostly nits.

Also maybe @SimonSapin wants to take a look.

SimonSapin commented on components/selectors/parser.rs:
```diff
@@ -1732,19 +1746,44 @@ pub mod tests {
-    fn parse<'i>(input: &'i str) -> Result<SelectorList<DummySelectorImpl>,
-                                           ParseError<'i, SelectorParseError<'i, ()>>> {
+    fn parse<'i>(input: &'i str)
+                 -> Result<SelectorList<DummySelectorImpl>, ParseError<'i, SelectorParseError<'i, ()>>>
```
No need to change this now, but next time you’d like to make whitespace-only changes please make them in a separate commit.

SimonSapin reviewed (commented):
Looks fine to me, with @emilio’s comment addressed.

jyc commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 1301bcdf108e8de76b8db670fcbbf330a128212e with merge c15702942a6259d1f80cf7fc4486dbdd0acec084...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5079)

jyc commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit 1301bcdf108e8de76b8db670fcbbf330a128212e with merge 2320d595c247a2fb7aa6b18ce679f9b10426f53a...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/7945), [arm32](http://build.servo.org/builders/arm32/builds/8167), [arm64](http://build.servo.org/builders/arm64/builds/8201), [linux-dev](http://build.servo.org/builders/linux-dev/builds/8287), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5048), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5081), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/8281), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/759), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/757), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/4984), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5042), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/760), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/765), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3525)
State: approved= try=True

emilio commented:
@bors-servo r+

Awesome, thanks!

bors-servo commented:
:pushpin: Commit 1301bcd has been approved by `emilio`

<!-- @bors-servo r=emilio 1301bcdf108e8de76b8db670fcbbf330a128212e -->

bors-servo commented:
:hourglass: Testing commit 1301bcdf108e8de76b8db670fcbbf330a128212e with merge 97023f18f34413d79b0c7e0b7d5cb3781868392f...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/7949), [arm32](http://build.servo.org/builders/arm32/builds/8171), [arm64](http://build.servo.org/builders/arm64/builds/8205), [linux-dev](http://build.servo.org/builders/linux-dev/builds/8291), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5052), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5085), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/8285), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/763), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/761), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/4988), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5046), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/764), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/769), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3529)
Approved by: emilio
Pushing 97023f18f34413d79b0c7e0b7d5cb3781868392f to master...
